### PR TITLE
*: Report tracing results to Jaeger

### PR DESCRIPTION
**Pull request by zhongzc**
Signed-off-by: zhongzc <zhongzc_arch@outlook.com>

### What problem does this PR solve?

Lack of a mechanism on TiKV to explore tracing results.

Related feature requests: pingcap/tidb#17693
Related TiDB side & usage: pingcap/tidb#19557
Roadmap: pingcap-incubator/tidb-dashboard#710

### What is changed and how it works?

Display tracing results directly via Jaeger UI.

What's Changed:

- A trait `tracing::Reporter` is introduced with its two implementations `JaegerReporter` and `NullReporter`:
  - `JaegerReporter` reports tracing results to a UDP agent asynchronously via a tokio runtime.
  - `NullReporter` drops all results passed to it.
- Some configurations are introduced for reporting tracing results:
  - `jaeger-agent`: Jaeger agent, commonly set to `"127.0.0.1:6831"`.
  - `num-threads`: the number of worker threads for reporting.
  - `duration-threshold`: the lower limit of duration of an individual procedure to report. A tracing result with a duration less than this value will be discarded.
  - `spans-max-length`: the upper limit of the number of spans of a tracing result to report. The excess spans will be discarded.
- A macro `requests_to_trace` to register requests to trace. If a request is not within this list (i.e. it's not intended to trace), even though lots of spans are left for the request in the codebase, we don't need to pay for it.

### Release note 
- Report tracing results to Jaeger

### CODE REVIEW

**Comment by breeswish** [comment]
Why do we have so long "collect metrics"?

**Comment by zhongzc** [comment]
> Cool! But I'm interested in why do we have so long "collect metrics"?

Me too. That's the reason why I traced and showed it deliberately. XD

**Comment by zhongzc** [comment]
It might be due to the warming-up. I re-run it multiple times to warm TikV up this time then got a more reasonable output.

File: src/server/service/kv.rs [hunk 1]
```diff
@@ -135,16 +139,45 @@ impl<T: RaftStoreRouter<RocksEngine> + 'static, E: Engine, L: LockManager> Servi
...
     ($fn_name: ident, $future_name: ident, $req_ty: ident, $resp_ty: ident) => {
         fn $fn_name(&mut self, ctx: RpcContext<'_>, req: $req_ty, sink: UnarySink<$resp_ty>) {
             if !check_common_name(self.security_mgr.cert_allowed_cn(), &ctx) {
                 return;
             }
+
             let begin_instant = Instant::now_coarse();
+
+            let (_root_span, collector) = trace_may_enable!($fn_name, !self.tracing_reporter.is_null());
+            tracing::property(tracing::Key::Foo, "Bar".to_string()); // FIXME: for demonstration
```
Comment (breeswish) [inline src/server/service/kv.rs:171]: fix?

File: src/storage/mod.rs [hunk 2]
```diff
@@ -770,6 +779,9 @@ impl<E: Engine, L: LockManager> Storage<E, L> {
                     let begin_instant = Instant::now_coarse();
                     let mut stats = Statistics::default();
                     let r = Self::raw_get_key_value(&snapshot, cf, key, &mut stats);
+
+                    let _g = minitrace::new_span(tracing::Event::TiKvCollectMetrics as u32);
```
Comment (breeswish) [inline src/storage/mod.rs:783]: how about:

```rust
{
   let _g = ...
}
```

to exclude Drop / send time of `r` and other things

Comment (zhongzc) [reply src/storage/mod.rs:783]: I don't really want to trace `TiKvCollectMetrics`. It's also for demonstration and will be removed if the rest is ok.

Comment (breeswish) [reply src/storage/mod.rs:783]: Got it

**Comment by zhongzc** [comment]
Screenshot from 2020-08-20 21-47-11
Trace async codes of channel model.

**Comment by zhongzc** [comment]
Support to trace raftstore:

Screenshot from 2020-10-23 10-51-30
Screenshot from 2020-10-23 10-52-09

**Comment by breeswish** [comment]
Should we close this PR for https://github.com/tikv/tikv/pull/8981?

**Pull request closed without merging**